docs/tutorials/streaming_analytics.md

```diff
@@ -0,0 +1,460 @@
+only useful if you are going to do your own custom windowing, with a custom Trigger. In many cases
+where this might seem useful you will be better off using a `ProcessFunction` as described [in
+another section]({{ site.baseurl }}{% link tutorials/event_driven.md %}#process-functions).
+
+### Window Functions
+
+You have three basic options for how to process the contents of your windows:
+
+1. as a batch, using a `ProcessWindowFunction` that will be passed an `Iterable` with the window's contents;
+1. incrementally, with a `ReduceFunction` or an `AggregateFunction` that is called as each event is assigned to the window;
+1. or with a combination of the two, wherein the pre-aggregated results of a `ReduceFunction` or an `AggregateFunction` are supplied to a `ProcessWindowFunction` when the window is triggered.
+
+Here are examples of approaches 1 and 3. Each implementation finds the peak value from each sensor
+in 1 minute event time windows, and producing a stream of Tuples containing `(key,
+end-of-window-timestamp, max_value)`.
+
+#### ProcessWindowFunction Example
+
+{% highlight java %}
+DataStream<SensorReading> input = ...
+
+input
+    .keyBy(x -> x.key)
+    .window(TumblingEventTimeWindows.of(Time.minutes(1)))
+    .process(new MyWastefulMax());
+
+public static class MyWastefulMax extends ProcessWindowFunction<
+        SensorReading,                  // input type
+        Tuple3<String, Long, Integer>,  // output type
+        String,                         // key type
+        TimeWindow> {                   // window type
+    
+    @Override
+    public void process(
+            String key,
+            Context context, 
+            Iterable<SensorReading> events,
+            Collector<Tuple3<String, Long, Integer>> out) {
+
+        int max = 0;
+        for (SensorReading event : events) {
+            if (event.value > max) max = event.value;
+        }
+        out.collect(new Tuple3<>(key, context.window().getEnd(), max));
+    }
+}
+{% endhighlight %}
+
+A couple of things to note in this implementation:
+
+* All of the events assigned to the window have to be buffered in keyed Flink state until the window
+  is triggered. This is potentially quite expensive.
+* Our ProcessWindowFunction is being passed a Context object from which contains information about
+  the window. Its interface looks like this:
+
+{% highlight java %}
+public abstract class Context implements java.io.Serializable {
+    public abstract W window();
+    
+    public abstract long currentProcessingTime();
+    public abstract long currentWatermark();
+
+    public abstract KeyedStateStore windowState();
+    public abstract KeyedStateStore globalState();
+}
+{% endhighlight %}
+
+`windowState` and `globalState` are places where you can store per-key, per-window, or global
+per-key information. This might be useful, for example, if you want to record something about the
```

Comment:
I often find this distinction (naming) a bit confusing. How about this to make it clear(er)?
```suggestion
`windowState` and `globalState` are places where you can store per-key, per-window, or global
per-key information for all windows of that key. This might be useful, for example, if you want to
record something about the
```